spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/outbound/HazelcastCacheWritingMessageHandler.java

```diff
@@ -16,64 +16,147 @@
+		this.extractPayload = extractPayload;
+	}
+
 	@Override
 	protected void onInit() throws Exception {
 		super.onInit();
-		HazelcastIntegrationDefinitionValidator.validateCacheTypeForCacheWritingMessageHandler(this.distributedObject);
 	}
 
 	@Override
-	protected void handleMessageInternal(Message<?> message) throws Exception {
+	protected void handleMessageInternal(final Message<?> message) throws Exception {
 		writeToCache(message);
 	}
 
 	@SuppressWarnings({ "unchecked", "rawtypes" })
-	private void writeToCache(Message<?> message) {
-		if (this.distributedObject instanceof IMap) {
-			((IMap<?, ?>) this.distributedObject).putAll((Map) message.getPayload());
+	private void writeToCache(final Message<?> message) {
+		DistributedObject distributedObject = getDistributedObject(message);
+		if (distributedObject instanceof IMap) {
+			((IMap) distributedObject).put(parseKeyExpression(message),
+					getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof MultiMap) {
+			((MultiMap) distributedObject).put(parseKeyExpression(message),
+					getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof ReplicatedMap) {
+			((ReplicatedMap) distributedObject).put(parseKeyExpression(message),
+					getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof IList) {
+			((IList) distributedObject).add(getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof ISet) {
+			((ISet) distributedObject).add(getPayloadOrMessage(message));
+		}
+		else if (distributedObject instanceof IQueue) {
+			((IQueue) distributedObject).add(getPayloadOrMessage(message));
 		}
-		else if (this.distributedObject instanceof IList) {
-			((IList<?>) this.distributedObject).addAll((List) message.getPayload());
+		else if (distributedObject instanceof ITopic) {
+			((ITopic) distributedObject).publish(getPayloadOrMessage(message));
 		}
-		else if (this.distributedObject instanceof ISet) {
-			((ISet<?>) this.distributedObject).addAll((Set) message.getPayload());
+	}
+
+	private DistributedObject getDistributedObject(final Message<?> message) {
+		if (this.distributedObject != null) {
+			return this.distributedObject;
+		}
+		else if (StringUtils.hasText(this.cacheExpression)) {
+			return parseCacheExpression(message);
+		}
+		else if (message.getHeaders().get(HazelcastHeaders.CACHE_NAME) != null) {
+			return this.getBeanFactory().getBean(
+					message.getHeaders().get(HazelcastHeaders.CACHE_NAME).toString(),
+					DistributedObject.class);
+		}
+		else {
+			throw new IllegalStateException("One of 'cache', 'cache-expression' and "
+					+ HazelcastHeaders.CACHE_NAME
+					+ " must be set for cache object definition.");
+		}
+	}
+
+	private DistributedObject parseCacheExpression(final Message<?> message) {
+		Object cacheName = parseExpression(this.cacheExpression, message);
+		return this.getBeanFactory().getBean(cacheName.toString(),
+				DistributedObject.class);
```

Comment:
I have the following concern :

I have implemented like this because the messages may need to be written to different cache objects by deciding at runtime. 

For example : ```cache-expression="headers['CACHE_HEADER']"```.
first message => distributedMap
second message => distributedList

However, as far as i see, we can not support it via ```cache-expression="@distributedList"``` and if the bean does not exist in context.